Client/Frontend/Browser/Favicons/FaviconFetcher.swift

```diff
@@ -0,0 +1,530 @@
+        }
+        fetchIcon { url, attributes in
+            DispatchQueue.main.async {
+                completion(url, attributes)
+            }
+        }
+    }
+    
+    // MARK: - Custom Icons
+    
+    /// Icon attributes for any custom icon overrides
+    ///
+    /// If the app does not contain a custom icon for the site provided `nil`
+    /// will be returned
+    var customIcon: FaviconAttributes? {
+        guard let folder = FileManager.default.getOrCreateFolder(name: NTPDownloader.faviconOverridesDirectory),
+            let baseDomain = url.baseDomain else {
+                return nil
+        }
+        let backgroundName = baseDomain + NTPDownloader.faviconOverridesBackgroundSuffix
+        let backgroundPath = folder.appendingPathComponent(backgroundName)
+        do {
+            let colorString = try String(contentsOf: backgroundPath)
+            let colorFromHex = UIColor(colorString: colorString)
+            
+            if FileManager.default.fileExists(atPath: folder.appendingPathComponent(baseDomain).path) {
+                let imagePath = folder.appendingPathComponent(baseDomain)
+                if let image = UIImage(contentsOfFile: imagePath.path) {
+                    return FaviconAttributes(
+                        image: image,
+                        backgroundColor: colorFromHex
+                    )
+                }
+                return nil
+            }
+        } catch {
+            return nil
+        }
+        return nil
+    }
+    
+    // MARK: - Bundled Icons
+    
+    /// Icon attributes for icons that are bundled in the app by default.
+    ///
+    /// If the app does not contain the icon for the site provided `nil` will be
+    /// returned
+    var bundledIcon: FaviconAttributes? {
+        // Problem: Sites like amazon exist with .ca/.de and many other tlds.
+        // Solution: They are stored in the default icons list as "amazon" instead of "amazon.com" this allows us to have favicons for every tld."
+        // Here, If the site is in the multiRegionDomain array look it up via its second level domain (amazon) instead of its baseDomain (amazon.com)
+        let hostName = url.hostSLD
+        var bundleIcon: (color: UIColor, url: String)?
+        if Self.multiRegionDomains.contains(hostName), let icon = Self.bundledIcons[hostName] {
+            bundleIcon = icon
+        } else if let name = url.baseDomain, let icon = Self.bundledIcons[name] {
+            bundleIcon = icon
+        }
+        guard let icon = bundleIcon, let image = UIImage(contentsOfFile: icon.url) else {
+            return nil
+        }
+        let scale = self.kind == .largeIcon ? 40.0 : 20.0
+        return FaviconAttributes(
+            image: image.createScaled(CGSize(width: scale, height: scale)),
+            backgroundColor: icon.color,
+            contentMode: .center
+        )
+    }
+    
+    private static let multiRegionDomains = ["craigslist", "google", "amazon"]
```

Comment:
@kylehickinson - You'll run into a different problem here where subdomains like:

1. keep.google.com
2. meet.google.com

have different favicons.

Reply:
Hmm so I think this might be OK, because these are only used for bundled icons and bundled icons are only used when there isn't a valid favicon for the given domain.  If a user visits meet.google.com the favicon will show correctly (on favorites it actually should show the regular google icon since this is the `apple-touch-icon-precomposed` it has when visiting from a mobile app: https://storage.googleapis.com/operating-anagram-8280/apple-touch-icon.png)